### PR TITLE
Remove kwargs in policy init

### LINKED ISSUE #947: Remove kwargs in policies init

### PR DESCRIPTION
**Pull request by MischaPanch**
Closes #947 

This removes all kwargs from all policy constructors. While doing that, I also improved several names and added a whole lot of TODOs.

## Functional changes:

1. Added possibility to pass None as `critic2` and `critic2_optim`. In fact, the default behavior then should cover the absolute majority of cases
2. Added a function called `clone_optimizer` as a temporary measure to support passing `critic2_optim=None` 

## Breaking changes:

1. `action_space` is no longer optional. In fact, it already was non-optional, as there was a ValueError in BasePolicy.init. So now several examples were fixed to reflect that
2. `reward_normalization` removed from DDPG and children. It was never allowed to pass it as `True` there, an error would have been raised in `compute_n_step_reward`. Now I removed it from the interface
3. renamed `critic1` and similar to `critic`, in order to have uniform interfaces. Note that the `critic` in DDPG was optional for the sole reason that child classes used `critic1`. I removed this optionality (DDPG can't do anything with `critic=None`)
4. Several renamings of fields (mostly private to public, so backwards compatible)

## Additional changes: 
1. Removed type and default declaration from docstring. This kind of duplication is really not necessary
2. Policy constructors are now only called using named arguments, not a fragile mixture of positional and named as before
5. Minor beautifications in typing and code 
6. Generally shortened docstrings and made them uniform across all policies (hopefully)

## Comment:

With these changes, several problems in tianshou's inheritance hierarchy become more apparent. I tried highlighting them for future work.

### CODE REVIEW

**Comment by MischaPanch** [comment]
@Trinkle23897 this is a next big step towards more robust and maintainable code, it will require your attention.

These changes required some serious, soul-crushing grinding on the whole code base :cry: . Kwargs are such an insidious antipattern, once it's there, it's really hard to remove.

I described all changes that could be somehow controversial above, the rest is just pushing args around and adjusting everything after.

@opcode81 The changes will probably slightly affect the high level interfaces, e.g., the removal of `reward_normalization` for DDPG and children. Should make it easier for you to directly see which configuration a policy can take

**Comment by MischaPanch** [comment]
Ah, one more thing @Trinkle23897 : I think it would be better not to squash the commits on this PR, so that eventual mistakes can be caught more easily. I tried to structure the commits in a sensible way, which should be much better for transparency than a single, huge commit over 2k lines of code

**Comment by Trinkle23897** [comment]
Actually, could you do the following two things before this PR?
1. Support mypy for all files, including examples and tests
2. Change function signature to 
```
def func(
  self,
  *,  # add this `*` to force use kwargs
  ...
) -> ...:
  pass
```

**Comment by MischaPanch** [comment]
Good point about the forcing of kwargs! Wish it'd been done earlier ^^

I'll do it in this PR and not in a separate one since it would create lot's of conflicts and duplicated work otherwise

> Actually, could you do the following two things before this PR?
> 
> 1. Support mypy for all files, including examples and tests
> 2. Change function signature to
> 
> ```
> def func(
>   self,
>   *,  # add this `*` to force use kwargs
>   ...
> ) -> ...:
>   pass
> ```

**Comment by MischaPanch** [comment]
@Trinkle23897 mypy on examples and tests finds a total of over 1000 typing violations. I would strongly prefer to handle that in a separate PR, maybe multiple ones.

I created #952 for that

**Comment by MischaPanch** [comment]
@Trinkle23897 this is ready for review/merge. Since it will create conflicts with almost any other PR, would be great to merge it soon :)

**Comment by Trinkle23897** [comment]
I'll take a carefully view

File: examples/atari/atari_c51.py
Comment (Trinkle23897) [inline examples/atari/atari_c51.py:91]: need to modify the following parts together:
https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/README.md?plain=1#L227
https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/docs/tutorials/dqn.rst?plain=1#L153
https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/docs/tutorials/tictactoe.rst?plain=1#L314
https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/docs/tutorials/tictactoe.rst?plain=1#L329

File: examples/offline/d4rl_cql.py
Comment (Trinkle23897) [inline examples/offline/d4rl_cql.py]: missing kwargs and action_space?

File: pyproject.toml
Comment (Trinkle23897) [inline pyproject.toml:105]: oh wow does it work now?

File: test/base/test_collector.py
Comment (Trinkle23897) [inline test/base/test_collector.py]: Won't action_space=None raise ValueError in BasePolicy? How come it currently passed all tests?

File: test/base/test_env_finite.py
Comment (Trinkle23897) [inline test/base/test_env_finite.py]: same here as above

File: test/offline/gather_pendulum_data.py
Comment (Trinkle23897) [inline test/offline/gather_pendulum_data.py]: rename critic1 to critic?

File: test/offline/test_bcq.py
Comment (Trinkle23897) [inline test/offline/test_bcq.py]: same as here?

File: test/offline/test_cql.py
Comment (Trinkle23897) [inline test/offline/test_cql.py]: same?

File: tianshou/data/buffer/her.py
Comment (Trinkle23897) [inline tianshou/data/buffer/her.py:18]: would you mind changing the docstring style to google one next time?

File: tianshou/policy/imitation/discrete_crr.py
Comment (Trinkle23897) [inline tianshou/policy/imitation/discrete_crr.py]: missing docstring? action_bound_method for example

File: tianshou/policy/imitation/td3_bc.py
Comment (Trinkle23897) [inline tianshou/policy/imitation/td3_bc.py]: or `None` for no bounding

File: tianshou/policy/modelfree/a2c.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/a2c.py:37]: missing `Only used if the action_space is continuous.`?

File: tianshou/policy/imitation/td3_bc.py
Comment (Trinkle23897) [reply tianshou/policy/imitation/td3_bc.py]: missing obsevation_space?

File: tianshou/policy/modelbased/psrl.py
Comment (Trinkle23897) [inline tianshou/policy/modelbased/psrl.py]: can actually set action_scaling=False and action_bound_method=None here since psrl can be only applied on discrete case

File: tianshou/policy/modelfree/discrete_sac.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/discrete_sac.py:36]: missing observation_space?

File: tianshou/policy/modelfree/pg.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/pg.py:51]: could you double-check if the rendered html doc has any formatting issues locally?

File: tianshou/policy/modelfree/redq.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/redq.py:125]: why not just replace `_is_auto_alpha` to `is_auto_alpha`?

File: tianshou/policy/modelfree/sac.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/sac.py:48]: missing observation_space?

File: tianshou/policy/modelfree/td3.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/td3.py]: same missing observation_space?

File: tianshou/utils/optim.py
Comment (Trinkle23897) [inline tianshou/utils/optim.py:38]: Can we use their parent class instead of hardcoding things here? Otherwise, if pytorch has an upgrade that changes some of them, it would silently fail on our side.

File: tianshou/utils/optim.py
Comment (Trinkle23897) [inline tianshou/utils/optim.py:43]: why do we need that?

File: tianshou/utils/optim.py
Comment (Trinkle23897) [inline tianshou/utils/optim.py:23]: this `optim_step` is not covered by any test

File: pyproject.toml
Comment (MischaPanch) [reply pyproject.toml:105]: Nope, but at least you can run `mypy tests` and `mypy examples` to find the 1k+ typing issues. `poe type-check`, CI and pre-commit still run `mypy tianshou`

File: test/base/test_collector.py
Comment (MischaPanch) [reply test/base/test_collector.py]: Yeah, I didn't understand that either, but also didn't bother going through the tests. Maybe this class is never instantiated?

File: tianshou/data/buffer/her.py
Comment (MischaPanch) [reply tianshou/data/buffer/her.py:18]: If there is a reliable docstring converter, it could be done. But why? Is there a large benefit in it?

This would need to adjust the sphinx build, and maybe also things like doc8 config

File: tianshou/policy/imitation/discrete_crr.py
Comment (MischaPanch) [reply tianshou/policy/imitation/discrete_crr.py]: I tried my best to keep the docstrings correct, but there are probably still several incomplete ones... Gonna address the ones you found

File: tianshou/policy/modelfree/redq.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/redq.py:125]: Because the user doesn't pass `is_auto_aplha`, it is inferred from the `alpha` input (if it's a tuple or a constant). So the user code should never be able to mutate it, hence the property

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:38]: I purposedly didn't use the base class because the clone_optimizer method might not work for user-defined optimizers

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:43]: Because I relaxed the interface to pass `critic2=None, critic2_optim=None`. The default behavior is then to copy `critic` and `critic_optim`. The latter cannot be copied, hence this function. Note that this default behavior should cover the majority of use cases, and I believe it covers all examples and tests, even though it's not used there atm.

Generally, I think we should bundle the nets and optimizers in a single abstraction, or let users pass optimizer factories instead. Currently, we always rely on the user to have a correct optimizer, i.e. we hope that `critic_optim` was indeed created with `critic.parameters()`. This is not verified at runtime, and if it doesn't hold, the algorithms won't work, of course.

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:23]: Yes, it slipped in with a previous merge, where I have forgotten to delete it. I use it on other branches in my fork, and plan to include tests for it and use it across policies (this code is duplicated many times over). Can we keep it for the time being? It is currently unused

File: test/base/test_collector.py
Comment (Trinkle23897) [reply test/base/test_collector.py]: could you add a breakpoint when executing test?

File: tianshou/utils/optim.py
Comment (Trinkle23897) [reply tianshou/utils/optim.py:23]: sure

**Comment by Trinkle23897** [comment]
Oh the dumb thing: https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/tianshou/policy/base.py#L107
missing `raise`

**Comment by MischaPanch** [comment]
> Oh the dumb thing:
> 
> https://github.com/thu-ml/tianshou/blob/6449a43261eeea82185e07dd659a79afff9556ab/tianshou/policy/base.py#L107
> 
> 
> missing `raise`

How did me and my IDE miss this :( .

I was busy lately but will continue on this PR tomorrow and address your comments then

File: examples/offline/d4rl_cql.py
Comment (MischaPanch) [reply examples/offline/d4rl_cql.py]: Done

File: test/base/test_env_finite.py
Comment (MischaPanch) [reply test/base/test_env_finite.py]: Done

File: test/base/test_collector.py
Comment (MischaPanch) [reply test/base/test_collector.py]: Done (added the raise and passed an action space here)

File: test/offline/gather_pendulum_data.py
Comment (MischaPanch) [reply test/offline/gather_pendulum_data.py]: Done, also removed explicit construction of second critic and optim, so the new default behavior is covered by tests

File: test/offline/test_bcq.py
Comment (MischaPanch) [reply test/offline/test_bcq.py]: Done, as above

File: test/offline/test_cql.py
Comment (MischaPanch) [reply test/offline/test_cql.py]: Done

File: tianshou/utils/optim.py
Comment (opcode81) [reply tianshou/utils/optim.py:43]: Indeed, I don't think this is needed (or elegant). The algorithm requires both critics and if the user wants to use the same type of critic twice, he can create two identical instances. This is more appropriate for a low-level interface than a perhaps contrived cloning mechanism.
In the high-level interface, which is currently being developed (#938), the user will be offered more convenient interfaces anyway (involving factories). 
For the more low-level interface, I would prefer to keep it simple (and true to the actual requirements) and rather remove this.

File: tianshou/policy/imitation/td3_bc.py
Comment (MischaPanch) [reply tianshou/policy/imitation/td3_bc.py]: Made docstrings for these params the same as in DDPG

File: tianshou/policy/modelbased/psrl.py
Comment (MischaPanch) [reply tianshou/policy/modelbased/psrl.py]: Done

File: tianshou/policy/modelfree/discrete_sac.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/discrete_sac.py:36]: Done

File: tianshou/policy/modelfree/sac.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/sac.py:48]: Done

File: tianshou/policy/modelfree/td3.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/td3.py]: Done

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:43]: I disagree, AFAIK in virtually all cases in tianshou the `net_2`, `critic2`, and `critic_optim2` are covered by the default I introduced here. This will also be true for the absolute majority of users, I don't see the point in forcing them to create a second net/critic/optim if there is such a sensible default behavior. Of course, they are still able to do so if desired.

In my opinion, the main problem manifested here is that the nets and their optimizers are passed separately, as I outlined above, requiring this hacky and non-elegant solution. I view it as temporary step, until the "bundling" of optimizers and nets is improved - either by new abstractions of by switching to passing optim factories (like already done in many other torch-based libraries)

Comment (MischaPanch) [reply tianshou/utils/optim.py:43]: I am also not aware of any recommendation from research about using different model architectures or learning rates for the dual-critic algorithms. There might be some exotic cases, of course

Comment (opcode81) [reply tianshou/utils/optim.py:43]: IMO, clarity and simplicity should be prioritized over ease of use in a low-level interface. The high-level interface will solve the usability issues, so there is not really a good reason to solve some of them twice (in different ways).
It is just clearer to require exactly what the algorithm actually requires in the interface than to have a default of `None` which needs to be explained (and requires additional code to be handled).

Comment (opcode81) [reply tianshou/utils/optim.py:43]: I would be more inclined to agree with your notion if there was no high-level interface in development, mind you.
If you continue down this path, the low-level interface will, more and more, become a high-level interface which focuses on usability rather than simplicity and clarity, introducing unnecessary complexity.

File: examples/atari/atari_c51.py
Comment (MischaPanch) [reply examples/atari/atari_c51.py:91]: Done

File: tianshou/utils/optim.py
Comment (opcode81) [reply tianshou/utils/optim.py:43]: There is value in maintaining a high degree of simplicity in the core (low-level) part of the library ([KISS](https://www.baeldung.com/cs/kiss-software-design-principle)).

Comment (MischaPanch) [reply tianshou/utils/optim.py:43]: I don't think this one change leads to a slippery slope ;). The policy interface in a way is a high-level interface, and I suspect that researchers will continue using it as main entry point for a while.

However, I do see how a bundling of nets and optimizers (or use of factories) in the future might increase complexity if not done with much care, so I'm not 100% sure either.

I think it's a general design decision here - should we put some weight on increasing the usability of the current interfaces, despite the coming addition of high level interfaces. My feeling is that researches will mainly rely on the current interfaces, while the new ones will be useful for newcomers and in applications. This is b/c researchers who develop algorithms will usually create a new policy by inheriting from existing ones. So in my opinion, potentially sacrificing a bit of explicitness for covering this stark default behavior is a positive development. I also view the fact that optimizers are passed explicitly (instead of through factories or abstractions) as technical debt, and not necessarily improving on simplicity of the core library. But I understand that it's not clear-cut

@Trinkle23897 your call to give the decisive vote.

File: tianshou/utils/optim.py
Comment (opcode81) [reply tianshou/utils/optim.py:38]: This already indicates that cloning the optimizer might not be the best approach in the first place.

File: tianshou/policy/modelfree/pg.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/pg.py:51]: I went over it, looks good to me, links are resolved

File: tianshou/policy/modelfree/a2c.py
Comment (MischaPanch) [reply tianshou/policy/modelfree/a2c.py:37]: Done

File: tianshou/policy/imitation/discrete_crr.py
Comment (MischaPanch) [reply tianshou/policy/imitation/discrete_crr.py]: Done

File: tianshou/utils/optim.py
Comment (Trinkle23897) [reply tianshou/utils/optim.py:43]: I'm more inclined toward @opcode81. I agree with @MischaPanch that we should make a better interface, and I think 
> I view it as temporary step, until the "bundling" of optimizers and nets is improved - either by new abstractions of by switching to passing optim factories (like already done in many other torch-based libraries)

is the right direction we should push forward. As an intermediate step, if we don't care about clean/elegant, then the current stuff is totally fine; however, if we do care about it in this intermediate step, I tend to revert the optimizer change in this file, for two reasons:
- we will refactor this whole interface anyway in the future
- it's confusing people if the algorithm uses two critics/optimizers but the entrypoint only passes one critic and optimizer (that's why at the beginning I made the entrypoint to explicitly pass two critics instead of one)

**Review by Trinkle23897 — APPROVED** [review]
Could you trigger another ci run to make sure everything looks good? If gpu test still fails, I'll take it anyway.

GPU test is flaky because it's in mainland China, which stays inside the GFW and needs "help" connecting to github stablely.

**Comment by MischaPanch** [comment]
> Could you trigger another ci run to make sure everything looks good? If gpu test still fails, I'll take it anyway.
> 
> GPU test is flaky because it's in mainland China, which stays inside the GFW and needs "help" connecting to github stablely.

CI tests passed on penultimate commit (except the GPU runner), and the last one was just an additional line in a docstring. Didn't seem worth retriggering the long tests over it.

@Trinkle23897 feel free to merge at your convenience, but I really recommend to add all commits and not squash them here. I enabled that option in the repo's settings

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:43]: Since atm it's clearly marked as intermediate step and not to be used directly, I think it's fine to keep it for the time being. We should continue the discussion in a separate setting, probably only after the first version of the high-level interfaces has been merged. @opcode81 is preparing a Wiki entry for it. I will create an issue linking to this thread, so we keep it in mind.

File: tianshou/utils/optim.py
Comment (MischaPanch) [reply tianshou/utils/optim.py:38]: See convo below, resolving

**Comment by Trinkle23897** [comment]
> I really recommend to add all commits and not squash them here

It’s hard to do bisect in that way. You can always first do bisect on master branch, and if you find this PR is causing the issue, you can checkout this branch and do another round of bisect.

**Comment by opcode81** [comment]
> > I really recommend to add all commits and not squash them here
> 
> It’s hard to do bisect in that way. You can always first do bisect on master branch, and if you find this PR is causing the issue, you can checkout this branch and do another round of bisect.

@Trinkle23897 it's not just about bisecting. But bisecting is rendered rather difficult if the original branch no longer exists (deleted upon merge) and isn't even in the same repository.

Importantly, with the original commits intact, 
 * you get much more information on the individual changes and can better trace why certain changes were made (e.g. when using *blame* on a file) - provided that each commit has a meaningful commit message.
 * you can choose to revert only individual commits later.

Also, my branch for the high-level API actually built on this branch, because it provided important foundations, so I will now need to rebuild my branch based on the new, rewritten history. It is not uncommon for branches to build upon each other, and rewriting the history along the way causes problems that could otherwise be avoided. If it had just been merged, I would only have needed to merge the updated master.

For these reasons, I generally advise against squashing. The only thing squashing has going for it is that it makes the history shorter/simpler. This is, in most cases, far less important than retaining more information and a seamless workflow when branches depend upon each other.

**Comment by Trinkle23897** [comment]
I'm sorry that caused some inconvenience for the development experience. In the future if I give a stamp, feel free to merge PR by yourself @MischaPanch (I believe you have the merge permission) in your own way.

**Comment by MischaPanch** [comment]
> I'm sorry that caused some inconvenience for the development experience. In the future if I give a stamp, feel free to merge PR by yourself @MischaPanch (I believe you have the merge permission) in your own way.

@Trinkle23897 I have fewer permissions than one might expect for a maintainer status. E.g., I'm not allowed to merge when the pipeline fails (as is the case most of the time b/c of the GPU runner), I also cannot change the way in which merge is performed (i.e., not squash) for forks. Tbh, I'm not even sure I can push to branches in tianshou, since so far GH rejected such pushes - but maybe something is wrong in my ssh settings.

I'm happy to take more responsibility for how things are merged, but I'd need the permissions for that. Maybe there are some settings that should be adjusted?

**Comment by Trinkle23897** [comment]
done, please check (previously you have "Maintain")
Screenshot 2023-10-09 at 12 06 38 PM

**Comment by MischaPanch** [comment]
Thank you! I promise to be responsible with the new powers :)

I will make a mini-PR now, supporting slightly older versions of gymnasium (currently tianshou requirements are in conflict with rllib, that's why), and a minor change to the readme. Since you're online, gonna ask you for a quick glance, but in future I might merge such minor changes directly without review, if you don't mind

**Comment by Trinkle23897** [comment]
Thanks! I'll make sure in the future I can give review for small-enough PR (diff < 200 lines) within 24h.

**Pull request merged**
